### ISSUE TITLE
Failure.parents has duplicate class

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4957]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)| @lvh reported|
|-|-|
|Trac ID|trac#4957|
|Type|defect|
|Created|2011-03-15 16:17:10Z|

In [#4928](https://github.com/twisted/twisted/issues/4928), amaury commented that twisted.python.failure.Failure.parents now contains the initial class twice. This is not a real correctness problem, but it's a bit stupid and inefficient.

Attachments:

* [4957-Failure.parents-has-duplicate-class.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/604/604c9f70880a080ae825505ad751fde6da520af8/85248bbb54f9a3c34f2d5f49582bdaf5cca080e9.patch) (418 bytes) - added by retenodus on 2011-06-29 00:33:19Z - Patch to remove duplicate class of Failure.parents 

<details><summary>Searchable metadata</summary>

```
trac-id__4957 4957
type__defect defect
reporter__lvh lvh
priority__lowest lowest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1300205830000000 1300205830000000
changetime__1310330443000000 1310330443000000
version__None None
owner__retenodus retenodus

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="retenodus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|retenodus set owner to retenodus<br>retenodus set status to `assigned`|
|-|-|

getmro already returns the type, so the line isn't necessary.

**Comment by twisted-trac** [comment]
|<img alt="retenodus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|retenodus removed owner<br>retenodus set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_3"></a>|@dustin set owner to retenodus|
|-|-|

This is very straightforward - presumably this error was introduced in the switch from allYourBase to getmro, where the former does not return the given class, and the latter does.  I'm tempted to say that this should have tests to verify that the the behavior has changed appropriately, but that seems like overkill here.

The existing tests continue to pass for me.  Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_4"></a>|@dustin commented|
|-|-|

Er, this also needs a topfiles addition - so do not merge yet!

**Comment by twisted-trac** [comment]
|<img alt="retenodus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|retenodus commented|
|-|-|

Euh, what topfile ?

I can't merge, the doc says that a core developper will merge it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

The [ReviewProcess#Newsfiles topfile] can be a .misc for this change I think.  I'll add it and apply the patch now.  Thanks for your work on this retenodus, and for the review dustin.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Ah, also, just to clarify one point.  I agree that we don't need a unit test here - because the patch is a pure deletion.  Generally it's okay to delete code without adding new tests, as long as some executing tests are executing the code being changed (if no tests had previously been running the if suite including the change, then a new test probably would have been required).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [32306]) Remove duplicate exception from `Failure.parents`

Author: retenodus
Reviewer: dustin
Fixes: [#4957](https://github.com/twisted/twisted/issues/4957)

When twisted/python/failure.py switched from `allYourBase` to `inspect.getmro`,
a duplicate of the type of `failure.value` was introduced into `failure.parents`.
Remove it.
